### PR TITLE
Discussion: Simplify the API for newcomers 

### PR DESCRIPTION
**Pull request by Hombre-x**
As @TonioGela and @armanbilge suggested, the API is bloated with difficult concepts such as `Resources` or context abstractions. 

To ensure that newcomers are not hindered in their encounter with these things, some deletion - refactoring - changing of the API was suggested to improve the experience with the library in general. 

This PR aims to open a discussion to merge some of these changes and further discuss if some more changes are required to accomplish this goal of enhancing the experience of new users.

What has been done so far: 

* Delete the `readStream` function to get rid of the `Stream` type and concepts.
* Squash the `WithCharset` method with their standalone variant (similar to [os-lib](https://github.com/com-lihaoyi/os-lib/blob/main/os/src/ReadWriteOps.scala#L228)) style.  
* Changed the signature of the `tempFile` and `tempDirectory` to not return a `Resource`.
* Refactored some tests to deal with the new API.

Whats left to discuss: 

The `As` variant of the three standalone methods was unchanged as I think we should talk before If I may get rid only of the typeclass use and instead add the codec as a dependency in the parameters, or, get rid of the methods entirely as well as with its example in the documentation.

### CODE REVIEW

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala [hunk 1]
```diff
@@ -705,14 +695,16 @@ object FilesOs {
   def size(path: Path): IO[Long] = files.size(path)
 
   /**
-   * Creates a temporary file and deletes it upon finalization of the returned
-   * resource.
+   * Creates a temporary file and deletes it at the end of the use of it.
    */
-  def tempFile: Resource[IO, Path] = files.tempFile(None, "", ".tmp", None)
+  def tempFile[A](use: Path => IO[A]): IO[A] =
+    IO.defer(files.createTempFile).bracket(use)(deleteIfExists(_).void)
```
Comment (TonioGela) [inline core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]:
```suggestion
    files.tempFile.use(use)
```

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala [hunk 2]
```diff
@@ -724,24 +716,31 @@ object FilesOs {
...
-  def tempFile(
+  def tempFile[A](
       dir: Option[Path],
       prefix: String,
       suffix: String,
       permissions: Permissions
-  ): Resource[IO, Path] = files.tempFile(dir, prefix, suffix, permissions.some)
+  )(use: Path => IO[A]): IO[A] = IO
+    .defer(files.createTempFile(dir, prefix, suffix, permissions.some))
+    .bracket(use)(deleteIfExists(_).void)
```
Comment (TonioGela) [inline core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]:
```suggestion
  )(use: Path => IO[A]): IO[A] = files.tempFile(dir, prefix, suffix, permissions.some).use(use)
```

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala [hunk 3]
```diff
@@ -724,24 +716,31 @@ object FilesOs {
...
-   * Creates a temporary directory and deletes it upon finalization of the
-   * returned resource.
+   * Creates a temporary directory and deletes it at the end of the use of it.
    */
-  def tempDirectory: Resource[IO, Path] = files.tempDirectory(None, "", None)
+  def tempDirectory[A](use: Path => IO[A]): IO[A] =
+    IO.defer(files.createTempDirectory)
+      .bracket(use)(deleteRecursively(_).recover {
+        case _: NoSuchFileException => ()
+      })
```
Comment (TonioGela) [inline core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]:
```suggestion
    files.tempDirectory.use(use)
```

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala [hunk 4]
```diff
@@ -751,13 +750,17 @@ object FilesOs {
...
+  def tempDirectory[A](
       dir: Option[Path],
       prefix: String,
       permissions: Permissions
-  ): Resource[IO, Path] = files.tempDirectory(dir, prefix, permissions.some)
+  )(use: Path => IO[A]): IO[A] = IO
+    .defer(files.createTempDirectory(dir, prefix, permissions.some))
+    .bracket(use)(deleteRecursively(_).recover { case _: NoSuchFileException =>
+      ()
+    })
```
Comment (TonioGela) [inline core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]:
```suggestion
  )(use: Path => IO[A]): IO[A] = files.tempDirectory(dir, prefix, permissions.some).use(use)

```

**Comment by TonioGela** [comment]
The simplification looks good, but I think you can do further, for example:

```diff
-def list(path: Path): Stream[IO, Path] = files.list(path)
+def list(path: Path): IO[List[Path]] = files.list(path).compile.toList

-  def open(path: Path, flags: Flags): Resource[IO, FileHandle[IO]] =
-   files.open(path, flags)

-  def readCursor(path: Path, flags: Flags): Resource[IO, ReadCursor[IO]] =
-   files.readCursor(path, flags)
```

> The `As` variant of the three standalone methods was unchanged as I think we should talk before If I may get rid only of the typeclass use and instead add the codec as a dependency in the parameters, or, get rid of the methods entirely as well as with its example in the documentation.

The thing is that I don't think any newcomer will ever use this methods, nor write a codec for something, BUT if we write a short tutorial on "how to implement your own saving format for your application" I might be in favor of keeping it there. It's not the main goal of this library, but we can discuss it.

Keep up with the good work. I'm willing to also see the site/doc changes that you're doing.
Cheers.

**Review by TonioGela — APPROVED** [review]

**Comment by zetashift** [comment]
For `read` weren't we going to default to `UTF-8`? 

EDIT: nvm see it correctly now!

**Review by zetashift — APPROVED** [review]

**Comment by TonioGela** [comment]
> For `read` weren't we going to default to `UTF-8`?
> 
> EDIT: nvm see it correctly now!

What have you seen? I'm asking it because I forgot what we decided about it, and forcing users to specify UTF-8 everywhere is a no-go.

**Comment by Hombre-x** [comment]
I guess he saw that the `read` method wasn't overloaded 🤔

**Comment by TonioGela** [comment]
> I guess he saw that the `read` method wasn't overloaded 🤔

Okay, but do we have just a single `read` method now? The one that doesn't default to utf-8? That's a no-go.

**Comment by zetashift** [comment]
> Okay, but do we have just a single `read` method now? The one that doesn't default to utf-8? That's a no-go.
We default to utf8:
https://github.com/davenverse/shellfish/blob/729108366a2bbf546ff4a89d50e1601319b3999f/core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala#L52

> I guess he saw that the `read` method wasn't overloaded 🤔
Yup, exactly what happened :P

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala [hunk 5]
```diff
@@ -705,14 +679,16 @@ object FilesOs {
   def size(path: Path): IO[Long] = files.size(path)
 
   /**
-   * Creates a temporary file and deletes it upon finalization of the returned
-   * resource.
+   * Creates a temporary file and deletes it at the end of the use of it.
    */
-  def tempFile: Resource[IO, Path] = files.tempFile(None, "", ".tmp", None)
+  def tempFile[A](use: Path => IO[A]): IO[A] =
```
Comment (lenguyenthanh) [inline core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]: bike-shedding: I found `tempFile` name is a bit unintuitive. I suggest adding `use` prefix for all of these functions, which reflects how we implement it as well `files.tempFile.use(use)`

I'm still not very happy with my suggestion, I think I need to see some practical usage of it to have a better suggestion. Other suggestions are welcome.

```suggestion
  def useTempFile[A](use: Path => IO[A]): IO[A] =
```

**Review by lenguyenthanh — APPROVED** [review]
I added a small name-shedding comment, which shouldn't block this pr.
Great works @Hombre-x and everyone!

File: core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala (hunk 5 above)
Comment (TonioGela) [reply core/src/main/scala/io/chrisdavenport/shellfish/FilesOs.scala]: `withTempFile`?

**Comment by Hombre-x** [comment]
Names changed 😀

**Review by TonioGela — APPROVED** [review]

**Pull request merged**